crates/hir-ty/src/infer.rs

```diff
@@ -333,6 +333,8 @@ pub struct InferenceResult {
     assoc_resolutions: FxHashMap<ExprOrPatId, AssocItemId>,
     pub diagnostics: Vec<InferenceDiagnostic>,
     pub type_of_expr: ArenaMap<ExprId, Ty>,
+    /// For each match expr, record diverging arm's expr.
+    pub diverging_arms: FxHashMap<ExprId, Vec<ExprId>>,
```

Comment:
I've tried to get diverging arms by looking types of arm expressions via `hir::semantics::Semantics::type_of_expr` but we coerce all match arm expressions and the whole match expression to the `Never` type when we encounter any diverging arm, so I couldn't find a way to distinguish diverging and non-diverging arms without recording this information.

Reply:
Does it help if you look at the unadjusted type? In an expression like

```rust
    match Some(0) {
        Some(_) => {}
        None => return
    }
```

`return` has type `!` but is adjusted to `()`, so the unadjusted type should provide the information you need

Reply:
Ah, I found the issue – your tests use `Option` but don't declare it. I'll push a fix.

Reply:
ohh 😅 great! Thanks for the fix.